File: game.of.life.v3/game.of.life.v3/Cell.cs
﻿namespace game.of.life.v3
{
    public class Cell
    {
        public Cell()
        { 
        }

        public Cell(int x, int y, CellState cellState = CellState.Dead)
            :this()
        {
            X = x;
            Y = y;
            State = cellState;
            NextState = CellState.Unknown;
        }

        public int X { get; private set; }

        public int Y { get; private set; }

        public CellState State { get; private set; }

        public CellState NextState { get; private set; }

        public bool IsAlive()
        {
            return State == CellState.Alive;
        }

        public void ComputeMutation(int aliveNeighboursCount)
        {
            if (State == CellState.Dead && aliveNeighboursCount == 3)
            {
                NextState = CellState.Alive;
            }

            if (State == CellState.Alive && (aliveNeighboursCount == 2 || aliveNeighboursCount == 3))
            {
                NextState = CellState.Alive;
            }

            if (State == CellState.Alive
                && (aliveNeighboursCount < 2 || aliveNeighboursCount >= 4))
            {
                NextState = CellState.Dead;
            }
        }

        private bool Equals(Cell other)
        {
            return X == other.X && Y == other.Y;
        }

        public override bool Equals(object obj)
        {
            if (ReferenceEquals(null, obj))
            {
                return false;
            }

            if (ReferenceEquals(this, obj))
            {
                return true;
            }

            return obj.GetType() == GetType() && Equals((Cell)obj);
        }

        public override int GetHashCode()
        {
            unchecked
            {
                return (X * 397) ^ Y;
            }
        }

        public override string ToString()
        {
            return string.Format("({0},{1})", X, Y);
        }

        public void CompleteMutation()
        {
            if (NextState != CellState.Unknown)
            {
                State = NextState;
                NextState = CellState.Unknown;
            }
        }
    }
}

File: game.of.life/game.of.life.ts.web/scripts/GoL/Logic/Cell.ts
/// <reference path="../../../typings/linq.d.ts"/>

'use strict';

namespace Gol.Logic {
	import CellState = GoL.Logic.CellState;

	export class Cell {
		private computeNextMutations: IComputeNextMutation[] =
			[
				new ShouldResurectNextMutation(),
				new ShouldStayAliveNextMutation(),
				new ShouldDieNextMutation()
			];

		public x: number;
		public y: number;
		public state: CellState;
		public nextState: CellState;

		constructor(x: number, y: number, state: CellState = CellState.Dead, nextState: CellState = CellState.Unknown) {
			this.x = x;
			this.y = y;
			this.state = state;
			this.nextState = nextState;
		}

		public static fromJsonArray(jsonCellsArray: string): Cell[] {
			return Enumerable.from(JSON.parse(jsonCellsArray))
				.cast<IJsonCell>().
				select((cell: IJsonCell) => new Cell(cell.x, cell.y))
				.toArray();
		}

		public isAlive(): boolean {
			return this.state === CellState.Alive;
		}

		public computeNextMutation(aliveNeighboursCount: number): void {
			const nextComputation = Enumerable.from(this.computeNextMutations)
				.firstOrDefault((m: IComputeNextMutation) => m.canComputeNextMutation(this, aliveNeighboursCount));
			
			if (nextComputation) {
				nextComputation.computeNextMutation(this);
			}
		}

		public completeMutation(): void {
			if (this.nextState === CellState.Unknown) {
				return;
			}

			this.state = this.nextState;
			this.nextState = CellState.Unknown;
		}

		public static equals(a: Cell, b: Cell): boolean {
			if (!a || !b) {
				return false;
			}

			return a.x === b.x && a.y === b.y;
		}
		
		public toString(): string
		{
			return `{"x":${this.x},"y":${this.y}}`;
		}
	}

	interface IComputeNextMutation {
		canComputeNextMutation(cell: Cell, aliveNeighboursCount: number): boolean;

		computeNextMutation(cell: Cell): void;
	}

	class ShouldResurectNextMutation implements IComputeNextMutation {
		public canComputeNextMutation(cell: Cell, aliveNeighboursCount: number): boolean {
			return cell.state === CellState.Dead && aliveNeighboursCount === 3;
		}

		public computeNextMutation(cell: Cell): void {
			cell.nextState = CellState.Alive;
		}
	}

	class ShouldStayAliveNextMutation implements IComputeNextMutation {
		public canComputeNextMutation(cell: Cell, aliveNeighboursCount: number): boolean {
			return cell.state === CellState.Alive && (aliveNeighboursCount === 2 || aliveNeighboursCount === 3);
		}

		public computeNextMutation(cell: Cell): void {
			cell.nextState = CellState.Alive;
		}
	}

	class ShouldDieNextMutation implements IComputeNextMutation {
		public canComputeNextMutation(cell: Cell, aliveNeighboursCount: number): boolean {
			return cell.state === CellState.Alive && (aliveNeighboursCount < 2 || aliveNeighboursCount >= 4);
		}

		public computeNextMutation(cells: Cell): void {
			cell.nextState = CellState.Dead;
		}
	}

	interface IJsonCell {
		x: number;
		y: number;
	}
}
